File: xrpld_netgen/libs/rippled.py
#!/usr/bin/env python
# coding: utf-8

import re
import os
import json
from datetime import datetime
from typing import Dict, Any, List  # noqa: F401

from xrpld_netgen.utils.misc import read_json
import hashlib

basedir = os.path.abspath(os.path.dirname(__file__))
parentdir = os.path.dirname(basedir)

def parse(value: str):
    if value == "no":
        return False
    else:
        return True


def get_feature_lines_from_path(path: str):
    with open(path, "r") as f:
        lines = f.readlines()
        return lines


def get_feature_lines_from_content(content: str):
    return content.decode('utf-8').splitlines()


def parse_rippled_amendments(lines: Any):
    amendments = {}
    for line in lines:
        if re.match(r"REGISTER_FEATURE", line) or re.match(r"REGISTER_FIX", line):
            amendment_name: str = ""
            if re.match(r"REGISTER_FIX", line):
                amendment_name = (
                    re.search("REGISTER_FIX\)?.*?\((.*?),", line).group(1) or 0
                )
            if re.match(r"REGISTER_FEATURE", line):
                amendment_name = (
                    re.search("REGISTER_FEATURE\((.*?),", line).group(1) or 0
                )
            supported = re.findall(r"Supported::(.*),", line)
            default_vote = re.findall(r"DefaultVote::(.*),", line)
            amendments[amendment_name] = {
                "supported": parse(supported[0] if supported else "no"),
                "default_vote": parse(default_vote[0] if default_vote else "no"),
            }
    return {
        k: hashlib.sha512(k.encode("utf-8")).digest().hex().upper()[:64]
        for (k, v) in amendments.items()
        if v["supported"] == True
    }


def convert_to_list_of_hashes(features):
    return list(features.values())


def update_amendments(features: Dict[str, Any], xrpl_protocol: str):
    # load the json string into a dictionary
    json_dict = read_json(f'{parentdir}/genesis.{xrpl_protocol}.json')

    new_amendments: List[str] = convert_to_list_of_hashes(features)

    # loop through the list of dictionaries in accountState
    for dct in json_dict['ledger']['accountState']:
        # check if the dictionary has a key called 'Amendments'
        if 'Amendments' in dct:
            # if it does, update it's value with the new amendments
            dct['Amendments'] = new_amendments

    return json_dict

File: src/libs/rippled.ts
import * as fs from 'fs';
import * as crypto from 'crypto';
import { downloadFileAtCommitOrTag } from './github';
import path from 'path';

interface Amendments {
  [key: string]: {
    supported: boolean;
    default_vote: boolean;
  };
}

const basedir = process.cwd()
const srcDir = path.join(basedir, 'src')

function readJson(filePath: string): any {
  const data = fs.readFileSync(filePath, 'utf-8');
  return JSON.parse(data);
}

function convertToListOfHashes(features: Record<string, any>): string[] {
  return Object.keys(features).map(key =>
    crypto.createHash('sha512').update(key).digest('hex').toUpperCase().slice(0, 64)
  );
}

export async  function updateAmendments(features: Record<string, any>, xrplProtocol: string) {
  const jsonText = fs.readFileSync(`${srcDir}/genesis.${xrplProtocol}.json`, 'utf-8')
  const jsonDict = JSON.parse(jsonText);

  const newAmendments: string[] = convertToListOfHashes(features);

  for (const dct of jsonDict['ledger']['accountState']) {
    if ('Amendments' in dct) {
      dct['Amendments'] = newAmendments;
    }
  }

  return jsonDict;
}

export function parseRippledAmendments(lines: string[]): Record<string, string> {
  const amendments: Amendments = {};
  const registerFeatureRegex = /REGISTER_FEATURE\((.*?),/;
  const registerFixRegex = /REGISTER_FIX\)?.*?\((.*?),/;
  const supportedRegex = /Supported::(.*),/;
  const defaultVoteRegex = /DefaultVote::(.*),/;

  for (const line of lines) {
    let amendmentName = '';
    if (registerFixRegex.test(line)) {
      amendmentName = registerFixRegex.exec(line)?.[1] || '';
    } else if (registerFeatureRegex.test(line)) {
      amendmentName = registerFeatureRegex.exec(line)?.[1] || '';
    }

    const supported = supportedRegex.exec(line)?.[1] || 'no';
    const defaultVote = defaultVoteRegex.exec(line)?.[1] || 'no';

    amendments[amendmentName] = {
      supported: supported === 'yes',
      default_vote: defaultVote === 'yes',
    };
  }

  return Object.fromEntries(
    Object.entries(amendments)
      .filter(([, v]) => v.supported)
      .map(([k]) => [k, crypto.createHash('sha512').update(k).digest('hex').toUpperCase().slice(0, 64)])
  );
}

export function getFeatureLinesFromPath(filePath: string): string[] {
  const content = fs.readFileSync(filePath, 'utf-8');
  return content.split('\n');
}

export function getFeatureLinesFromContent(content: string): string[] {
  return content.split('\n');
}
